File: ox_common/android/src/main/java/com/ox/ox_common/OXCommonPlugin.java
package com.ox.ox_common;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.provider.MediaStore;
import android.util.Log;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.Nullable;
import androidx.core.content.FileProvider;

import com.ox.ox_common.activitys.PermissionActivity;
import com.ox.ox_common.activitys.SelectPicsActivity;
import com.ox.ox_common.provides.CustomAnalyzeCallback;
import com.ox.ox_common.utils.ClipboardHelper;
import com.uuzuche.lib_zxing.activity.CodeUtils;

import java.io.File;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.embedding.android.FlutterFragmentActivity;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.ActivityResultListener;

/**
 * Title: OXCommonPlugin
 * Description: TODO(Fill in by oneself)
 * Copyright: Copyright (c) 2022
 * Company:  0xchat Teachnology
 * CreateTime: 2023/12/21 16:46
 *
 * @author Michael
 * @since JDK1.8
 */

public class OXCommonPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware {
    private MethodChannel channel;
    private Context mContext;
    private final String TAG = "OXCommonPlugin";

    private final int SELECT = 601;
    private final int READ_IMAGE = 603;

    private Result mResult;

    private FlutterFragmentActivity mActivity;

    private ActivityResultLauncher<String> mGetContent;
    private ActivityResultLauncher<String[]> requestPermissionLauncher;

    @Override
    public void onAttachedToEngine(FlutterPluginBinding flutterPluginBinding) {
        mContext = flutterPluginBinding.getApplicationContext();
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "ox_common");
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        mResult = result;
        switch (call.method) {
            case "getDatabaseFilePath":
                String dbName = call.argument("dbName");
                String databasefile = getDatabaseFilePath(dbName);
                if (mResult != null) {
                    mResult.success(databasefile);
                    mResult = null;
                }
                break;
            case "scan_path":
                String path = call.argument("path");
                CodeUtils.AnalyzeCallback analyzeCallback = new CustomAnalyzeCallback(result, mActivity.getIntent());
                CodeUtils.analyzeBitmap(path, analyzeCallback);
                break;
            case "getPickerPaths":
                getPickPaths(call);
                break;
            case "backToDesktop":
                if (mResult != null) {
                    mResult.success(true);
                    mResult = null;
                    mActivity.moveTaskToBack(false);
                }
                break;
            case "getPlatformVersion":
                if (mResult != null) {
                    mResult.success("Android ${android.os.Build.VERSION.RELEASE}");
                    mResult = null;
                }
                break;
            case "callSysShare":
                String filePath = call.argument("filePath");
                goSysShare(filePath);
                break;
            case "select34MediaFilePaths":
                int type = call.argument("type");
                select34MediaFilePaths(type);
                break;
            case "request34MediaPermission":
                int mediaType = call.argument("type");
                request34MediaPermission(mediaType);
                break;
            case "hasImages":
                if (mResult != null) {
                    boolean hasImages = ClipboardHelper.hasImages(mContext);
                    mResult.success(hasImages);
                    mResult = null;
                }
                break;
            case "getImages":
                if (mResult != null) {
                    List<String> imagePaths = ClipboardHelper.getImages(mContext);
                    mResult.success(imagePaths);
                    mResult = null;
                }
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    @Override
    public void onDetachedFromEngine(FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    @Override
    public void onDetachedFromActivity() {
    }

    @Override
    public void onReattachedToActivityForConfigChanges(ActivityPluginBinding binding) {
    }

    @Override
    public void onAttachedToActivity(ActivityPluginBinding binding) {
        Activity activity = binding.getActivity();
        mActivity = (FlutterFragmentActivity) activity;
        initializeActivityResultLauncher();
        binding.addActivityResultListener(new ActivityResultListener() {
            @Override
            public boolean onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
                if (resultCode != Activity.RESULT_OK) {
                    return false;
                }
                if (requestCode == SELECT) {
                    List<Map<String, String>> paths = (List<Map<String, String>>) data.getSerializableExtra(SelectPicsActivity.COMPRESS_PATHS);
                    if (mResult != null) {
                        mResult.success(paths);
                        mResult = null;
                    }
                } else if (requestCode == READ_IMAGE) {
                    Intent intent1 = new Intent(mActivity, SelectPicsActivity.class);
                    intent1.putExtras(data);
                    mActivity.startActivityForResult(intent1, SELECT);
                }
                return false;
            }
        });
    }

    private void initializeActivityResultLauncher() {
        mGetContent = mActivity.registerForActivityResult(new ActivityResultContracts.GetMultipleContents(),
                uris -> {
//                    Log.d("Michael", "mGetContent----uris ="+uris.toString());
                    List<String> filePaths = urisToFileList(uris);
//                    Log.d("Michael", "mGetContent----filePaths ="+filePaths);
                    if (mResult != null) {
                        mResult.success(filePaths);
                        mResult = null;
                    }
                });
        requestPermissionLauncher = mActivity.registerForActivityResult(
                new ActivityResultContracts.RequestMultiplePermissions(), permissions -> {
                    Boolean readMediaImagesGranted = permissions.getOrDefault(Manifest.permission.READ_MEDIA_IMAGES, false);
                    Boolean readMediaVideoGranted = permissions.getOrDefault(Manifest.permission.READ_MEDIA_VIDEO, false);
                    Boolean readMediaVisualUserSelectedGranted = permissions.getOrDefault(Manifest.permission.READ_MEDIA_VISUAL_USER_SELECTED, false);
                    Map<String, Boolean> mediaGranteds = new HashMap<>();
                    mediaGranteds.put("READ_MEDIA_IMAGES", readMediaImagesGranted);
                    mediaGranteds.put("READ_MEDIA_VIDEO", readMediaVideoGranted);
                    mediaGranteds.put("READ_MEDIA_VISUAL_USER_SELECTED", readMediaVisualUserSelectedGranted);
//                    Log.d("oxcommon", "requestPermissionLauncher---------readMediaImagesGranted ="+readMediaImagesGranted+"；readMediaVisualUserSelectedGranted ="+readMediaVisualUserSelectedGranted + "; readMediaVideoGranted ="+readMediaVideoGranted);
                    if (mResult != null) {
                        mResult.success(mediaGranteds);
                        mResult = null;
                    }
                });
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
    }

    private String getDatabaseFilePath(String dbName) {
        return mActivity.getDatabasePath(dbName).getPath();
    }

    private void select34MediaFilePaths(int type) {
        String input = "image/*";
        if (type == 1) {
            input = "image/*";
        } else if (type == 2) {
            input = "video/*";
        }
        mGetContent.launch(input);
    }
    private void request34MediaPermission(int type) {
        ///type: 1 - image, 2 - video
        if (Build.VERSION.SDK_INT >= 34) {
            String[] permissions = null;
            if (type == 1) {
                permissions = new String[]{
                        Manifest.permission.READ_MEDIA_IMAGES,
                        Manifest.permission.READ_MEDIA_VISUAL_USER_SELECTED
                };
            } else {
                permissions = new String[]{
                        Manifest.permission.READ_MEDIA_VIDEO,
                        Manifest.permission.READ_MEDIA_VISUAL_USER_SELECTED
                };
            }
            requestPermissionLauncher.launch(permissions);
        } else {
            if (mResult != null) {
                mResult.success(false);//Unsupported Android version
                mResult = null;
            }
        }
    }

    private void getPickPaths(MethodCall call){
        String galleryMode = call.argument("galleryMode");
        boolean showGif = call.argument("showGif");
        Map<String, Number> uiColor = call.argument("uiColor");
        Number selectCount = call.argument("selectCount");
        boolean showCamera = call.argument("showCamera");
        boolean enableCrop = call.argument("enableCrop");
        Number width = call.argument("width");
        Number height = call.argument("height");
        Number compressSize = call.argument("compressSize");
        String cameraMimeType = call.argument("cameraMimeType");
        Number videoRecordMaxSecond = call.argument("videoRecordMaxSecond");
        Number videoRecordMinSecond = call.argument("videoRecordMinSecond");
        Number videoSelectMaxSecond = call.argument("videoSelectMaxSecond");
        Number videoSelectMinSecond = call.argument("videoSelectMinSecond");
        String language = call.argument("language");

        Intent intent = new Intent();
        intent.putExtra(SelectPicsActivity.GALLERY_MODE, galleryMode);
        intent.putExtra(SelectPicsActivity.UI_COLOR, (Serializable) uiColor);
        intent.putExtra(SelectPicsActivity.SELECT_COUNT, selectCount);
        intent.putExtra(SelectPicsActivity.SHOW_GIF, showGif);
        intent.putExtra(SelectPicsActivity.SHOW_CAMERA, showCamera);
        intent.putExtra(SelectPicsActivity.ENABLE_CROP, enableCrop);
        intent.putExtra(SelectPicsActivity.WIDTH, width);
        intent.putExtra(SelectPicsActivity.HEIGHT, height);
        intent.putExtra(SelectPicsActivity.COMPRESS_SIZE, compressSize);
        intent.putExtra(SelectPicsActivity.CAMERA_MIME_TYPE, cameraMimeType);
        intent.putExtra(SelectPicsActivity.VIDEO_RECORD_MAX_SECOND, videoRecordMaxSecond);
        intent.putExtra(SelectPicsActivity.VIDEO_RECORD_MIN_SECOND, videoRecordMinSecond);
        intent.putExtra(SelectPicsActivity.VIDEO_SELECT_MAX_SECOND, videoSelectMaxSecond);
        intent.putExtra(SelectPicsActivity.VIDEO_SELECT_MIN_SECOND, videoSelectMinSecond);
        intent.putExtra(SelectPicsActivity.LANGUAGE, language);

        if (cameraMimeType != null) {
            intent.putExtra(PermissionActivity.PERMISSIONS, new String[]{Manifest.permission.CAMERA}
            );
            intent.setClass(mContext, PermissionActivity.class);
            mActivity.startActivityForResult(intent, READ_IMAGE);
        } else {
            intent.setClass(mContext, SelectPicsActivity.class);
            mActivity.startActivityForResult(intent, SELECT);
        }
    }

    private void goSysShare(String filePath) {
        Uri shareFileURI = null;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            shareFileURI = FileProvider.getUriForFile(mActivity,
                    mActivity.getPackageName() + ".fileprovider",
                    new File(filePath));
        } else {
            shareFileURI = Uri.fromFile(new File(filePath));
        }
        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        if (shareFileURI != null) {
            shareIntent.putExtra(Intent.EXTRA_STREAM, shareFileURI);
            shareIntent.setType("image/*");
            // Use sms_body to get text when the user selects SMS
            shareIntent.putExtra("sms_body", "");
        } else {
            shareIntent.setType("text/plain");
        }
        shareIntent.putExtra(Intent.EXTRA_TEXT, "");
        // Customize the title of the selection box
        mActivity.startActivity(Intent.createChooser(shareIntent, "Share"));
    }

    private List<String> urisToFileList(List<Uri> uris) {
        List<String> filePaths = new ArrayList<>();
        for (Uri uri : uris) {
            String path = getPathFromUri(mActivity, uri);
            if (path != null) {
                filePaths.add(path);
            }
        }
        return filePaths;
    }

    private String getPathFromUri(Context context, Uri uri) {
        String result = null;
        if ("content".equalsIgnoreCase(uri.getScheme())) {
            String[] projection = { MediaStore.Images.Media.DATA };
            Cursor cursor = null;
            try {
                cursor = context.getContentResolver().query(uri, projection, null, null, null);
                if (cursor != null && cursor.moveToFirst()) {
                    int column_index = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
                    result = cursor.getString(column_index);
                }
            } finally {
                if (cursor != null) {
                    cursor.close();
                }
            }
        }
        return result;
    }
}

File: ox_common/android/src/main/kotlin/com/ox/ox_common/OXCommonPlugin.kt
package com.ox.ox_common

import android.Manifest
import android.app.Activity
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import android.util.Log
import androidx.core.content.FileProvider
import com.ox.ox_common.activitys.PermissionActivity
import com.ox.ox_common.activitys.SelectPicsActivity
import com.ox.ox_common.provides.CustomAnalyzeCallback
import com.ox.ox_common.utils.BitmapUtils
import com.ox.ox_common.utils.FileTool
import com.ox.ox_common.utils.LocalConst
import com.ox.ox_common.utils.LocalTools
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.ActivityResultListener
import java.io.File
import java.io.Serializable
import com.uuzuche.lib_zxing.activity.CodeUtils


/** OXCommonPlugin */
class OXCommonPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var mContext: Context
    private val TAG: String = "OXCommonPlugin";

    private val SELECT = 601
    private val READ_IMAGE = 603

    private var mResult: Result? = null
    private var mIsNeedCrop = false
    private var _tempImageFileLocation: String? = null
    private var _mCropImgPath: String? = null
    private var _mFileName: String? = null

    private lateinit var mActivity: Activity

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        mContext = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "ox_common")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.hasArgument("isNeedTailor")) {
            mIsNeedCrop = call.argument<Boolean>("isNeedTailor")!!
        }
        mResult = result
        _tempImageFileLocation = null
        when (call.method) {
            "scan_path" -> {
                val path = call.argument<String>("path")
                val analyzeCallback: CodeUtils.AnalyzeCallback =
                    CustomAnalyzeCallback(result, mActivity.getIntent())
                CodeUtils.analyzeBitmap(path, analyzeCallback)
            }
            "getPickerPaths" -> {
                val galleryMode: String? = call.argument("galleryMode")
                val showGif: Boolean? = call.argument("showGif")
                val uiColor: Map<String, Number>? = call.argument("uiColor")
                val selectCount: Number? = call.argument("selectCount")
                val showCamera: Boolean? = call.argument("showCamera")
                val enableCrop: Boolean? = call.argument("enableCrop")
                val width: Number? = call.argument("width")
                val height: Number? = call.argument("height")
                val compressSize: Number? = call.argument("compressSize")
                val cameraMimeType: String? = call.argument("cameraMimeType")
                val videoRecordMaxSecond: Number? = call.argument("videoRecordMaxSecond")
                val videoRecordMinSecond: Number? = call.argument("videoRecordMinSecond")
                val videoSelectMaxSecond: Number? = call.argument("videoSelectMaxSecond")
                val videoSelectMinSecond: Number? = call.argument("videoSelectMinSecond")
                val language: String? = call.argument("language")

                val intent = Intent()

                intent.putExtra(SelectPicsActivity.GALLERY_MODE, galleryMode)
                intent.putExtra(SelectPicsActivity.UI_COLOR, uiColor as Serializable)
                intent.putExtra(SelectPicsActivity.SELECT_COUNT, selectCount)
                intent.putExtra(SelectPicsActivity.SHOW_GIF, showGif)
                intent.putExtra(SelectPicsActivity.SHOW_CAMERA, showCamera)
                intent.putExtra(SelectPicsActivity.ENABLE_CROP, enableCrop)
                intent.putExtra(SelectPicsActivity.WIDTH, width)
                intent.putExtra(SelectPicsActivity.HEIGHT, height)
                intent.putExtra(SelectPicsActivity.COMPRESS_SIZE, compressSize)
                //直接调用拍照或拍视频时有效
                //直接调用拍照或拍视频时有效
                intent.putExtra(SelectPicsActivity.CAMERA_MIME_TYPE, cameraMimeType)
                intent.putExtra(SelectPicsActivity.VIDEO_RECORD_MAX_SECOND, videoRecordMaxSecond)
                intent.putExtra(SelectPicsActivity.VIDEO_RECORD_MIN_SECOND, videoRecordMinSecond)
                intent.putExtra(SelectPicsActivity.VIDEO_SELECT_MAX_SECOND, videoSelectMaxSecond)
                intent.putExtra(SelectPicsActivity.VIDEO_SELECT_MIN_SECOND, videoSelectMinSecond)
                intent.putExtra(SelectPicsActivity.LANGUAGE, language)
                if (cameraMimeType != null) {
                    intent.putExtra(
                        PermissionActivity.PERMISSIONS,
                        arrayOf<String>(Manifest.permission.CAMERA)
                    )
                    intent.setClass(mContext, PermissionActivity::class.java)
                    mActivity.startActivityForResult(
                        intent,
                        READ_IMAGE
                    )
                } else {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                        intent.putExtra(
                            PermissionActivity.PERMISSIONS, arrayOf<String>(
                                Manifest.permission.READ_MEDIA_IMAGES,
                                Manifest.permission.READ_MEDIA_VIDEO
                            )
                        )
                        intent.setClass(mContext, PermissionActivity::class.java)
                        mActivity.startActivityForResult(
                            intent,
                            READ_IMAGE
                        )
                    } else {
                        intent.setClass(mContext, SelectPicsActivity::class.java)
                        mActivity.startActivityForResult(
                            intent,
                            SELECT
                        )
                    }
                }
            }
            "backToDesktop" -> {
                if (mResult != null) {
                    mResult!!.success(true)
                    mActivity.moveTaskToBack(false)
                }
            }
            "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
            "getDeviceId" -> result.success(LocalTools.getAndroidId(mContext))
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onDetachedFromActivity() {
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        mActivity = binding.activity


        binding.addActivityResultListener(ActivityResultListener { requestCode, resultCode, data ->
            if (resultCode != Activity.RESULT_OK) {
                return@ActivityResultListener false
            } else {
                if (requestCode == SELECT) {
                    val paths =
                        data?.getSerializableExtra(SelectPicsActivity.COMPRESS_PATHS) as List<Map<String, String>>
                    mResult?.success(paths)
                    return@ActivityResultListener true
                } else if (requestCode == READ_IMAGE) {
                    if (resultCode == Activity.RESULT_OK) {
                        var intent1 = Intent(mActivity, SelectPicsActivity::class.java)
                        intent1.putExtras(data!!)
                        mActivity.startActivityForResult(
                            intent1,
                            SELECT
                        )
                    }
                }
                false
            }
        })
    }

    override fun onDetachedFromActivityForConfigChanges() {

    }

    private fun processResult(content: String) {
        if (mResult == null) return
        mResult!!.success(content)
        mResult = null
    }


    fun takePhoto(activity: Activity, requestCode: Int, uriToBeSave: Uri?) {
        if (uriToBeSave == null) {
            Log.e(TAG, "uriToBeSave is null!")
            return
        }
        val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val photoURI = FileProvider.getUriForFile(mActivity, mActivity.getPackageName().toString() + ".fileprovider",
                    File(_tempImageFileLocation))
            intent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI)
        } else {
            intent.putExtra(MediaStore.EXTRA_OUTPUT, uriToBeSave)
        }
        intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
        activity.startActivityForResult(intent, requestCode)
    }

    fun takeVideo(activity: Activity, requestCode: Int, uriToBeSave: Uri?) {
        val takeVideoIntent = Intent(MediaStore.ACTION_VIDEO_CAPTURE)
        if (takeVideoIntent.resolveActivity(mActivity!!.packageManager) != null) {
            takeVideoIntent.putExtra("camerasensortype", 2) // Invoke the front camera
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                val videoURI = FileProvider.getUriForFile(mActivity, mActivity.getPackageName().toString() + ".fileprovider",
                        File(_tempImageFileLocation))
                takeVideoIntent.putExtra(MediaStore.EXTRA_OUTPUT, videoURI)
            } else {
                takeVideoIntent.putExtra(MediaStore.EXTRA_OUTPUT, getTempImageFileUri(".mp4"))
            }
            mActivity!!.startActivityForResult(takeVideoIntent, requestCode)
        }
    }

    fun choosePhoto(activity: Activity, requestCode: Int) {
        val intent = Intent()
        if (Build.VERSION.SDK_INT >= 30) {
            intent.action = Intent.ACTION_OPEN_DOCUMENT
            intent.addCategory(Intent.CATEGORY_OPENABLE)
        } else {
            intent.action = Intent.ACTION_GET_CONTENT
        }
        intent.type = "image/*"
        activity.startActivityForResult(intent, requestCode)
    }

    private fun getTempImageFileUri(fileSuffix: String): Uri? {
        val tempImageFileLocation = getTempImageFileLocation(fileSuffix)
        return if (tempImageFileLocation != null) {
            Uri.parse("file://$tempImageFileLocation")
        } else null
    }


    private fun getTempImageFileLocation(fileSuffix: String): String? {
        try {
            if (_tempImageFileLocation == null) {
                val avatarTempDirStr = getPicSavedDir()
                val avatarTempDir = File(avatarTempDirStr)
                if (avatarTempDir != null) {
                    // Create the directory if it doesn't exist.
                    if (!avatarTempDir.exists()) avatarTempDir.mkdirs()

                    // Temporary file name
                    _mFileName = System.currentTimeMillis().toString() + fileSuffix
                    _tempImageFileLocation = avatarTempDir.absolutePath + "/" + _mFileName
                    if (Build.VERSION.SDK_INT >= 30) {
                        val tempFile = File(_tempImageFileLocation)
                        if (!tempFile.exists()) {
                            if (tempFile.createNewFile()) {
                                _tempImageFileLocation = tempFile.absolutePath
                            }
                        }
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "【Pic】Encountered an error while trying to read the temporary storage path for local user images," + e.message, e)
        }
        Log.d(TAG, "【Pic】Currently retrieving the temporary storage path for local user images: $_tempImageFileLocation")
        return _tempImageFileLocation
    }

    private fun getPicSavedDir(): String? {
        var dir: String? = null
        dir = if (Environment.MEDIA_MOUNTED == Environment.getExternalStorageState() || !Environment.isExternalStorageRemovable()) {
            if (mIsNeedCrop && Build.VERSION.SDK_INT >= 30) {
                Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).absolutePath
            } else {
                mActivity!!.getExternalFilesDir(LocalConst.DIR_YLNEW_APP_FILES_PIC_DIR)!!.absolutePath
            }
        } else {
            //External storage is not available
            mActivity!!.filesDir.absolutePath + LocalConst.DIR_YLNEW_APP_FILES_PIC_DIR
        }
        //        LogUtils.e("dir = " + dir);
        return dir
    }

    /**
     * Initiate photo cropping
     */
    private fun startPhotoCrop(srcUri: Uri, cropUri: Uri, outputX: Int, outputY: Int, requestCode: Int) {
        val intent = Intent("com.android.camera.action.CROP")
        intent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION
        //Set source URI
        intent.setDataAndType(srcUri, "image/*")
        intent.putExtra("crop", "true")
        intent.putExtra("aspectX", 1)
        intent.putExtra("aspectY", 1)
        //        intent.putExtra("outputX", outputX);
//        intent.putExtra("outputY", outputY);
        intent.putExtra("scale", true)
        //        //Set image format
//        intent.putExtra("outputFormat", Bitmap.CompressFormat.JPEG.toString());
        intent.putExtra("return-data", false) //No need to return data to avoid exceptions due to large images
        intent.putExtra("noFaceDetection", true) // no face detection
        //Set destination URI
        intent.putExtra(MediaStore.EXTRA_OUTPUT, cropUri)
        mActivity!!.startActivityForResult(intent, requestCode)
    }

    private fun getSDK30PictureUri(): Uri? {
        val pictureDirectory = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
        if (!pictureDirectory.exists()) {
            pictureDirectory.mkdirs()
        }
        val imgFile = File(pictureDirectory.absolutePath + File.separator + _mFileName)
        try {
            if (!imgFile.exists()) {
                imgFile.createNewFile()
            }
        } catch (e: Exception) {
            Log.e("ox_debug", e.message, e)
        }
        _mCropImgPath = imgFile.absolutePath
        // Insert the file using the MediaStore API to obtain the URI where the system crop should be saved (since the app doesn't have permission to access public storage and needs to use the MediaStore API for operations)
        val values = ContentValues()
        values.put(MediaStore.Images.Media.DATA, _mCropImgPath)
        values.put(MediaStore.Images.Media.DISPLAY_NAME, _mFileName)
        values.put(MediaStore.Images.Media.MIME_TYPE, "image/*")
        return mActivity!!.contentResolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, values)
    }

    private fun getCompressionImg(filePath: String, quality: Int) {
        var bitmap: Bitmap? = null
        try {
            val opts = BitmapFactory.Options()
            opts.inJustDecodeBounds = false
            bitmap = BitmapUtils.loadLocalBitmap(filePath, opts)
        } catch (e: Exception) {
            Log.e("ox_debug", e.message, e)
        }
        if (bitmap == null) mResult!!.success(null)
        val path: String = FileTool.getFileSavedDir(mActivity, LocalConst.DIR_YLNEW_APP_FILES_PIC_DIR).toString() + "/" + System.currentTimeMillis() + ".png"
        val isSaveSucc: Boolean = FileTool.saveBitmap(bitmap, File(path), Bitmap.CompressFormat.JPEG, quality)
        if (isSaveSucc) {
            mResult!!.success(path)
        } else {
            mResult!!.success(null)
        }
    }

    private fun goSysShare(filePath: String) {
        var shareFileURI: Uri? = null
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            shareFileURI = FileProvider.getUriForFile(mActivity, mActivity.getPackageName().toString() + ".fileprovider",
                    File(filePath))
        } else {
            shareFileURI = Uri.fromFile(File(filePath))
        }
        var shareIntent: Intent = Intent(Intent.ACTION_SEND);
        if (shareFileURI != null) {
            shareIntent.putExtra(Intent.EXTRA_STREAM, shareFileURI)
            shareIntent.type = "image/*"
            //Use sms_body to get text when the user selects SMS
            shareIntent.putExtra("sms_body", "")
        } else {
            shareIntent.type = "text/plain"
        }
        shareIntent.putExtra(Intent.EXTRA_TEXT, "")
        //Customize the title of the selection box
        mActivity.startActivity(Intent.createChooser(shareIntent, "Share"))
    }

}
